Decouple release reconciliation from chart source sync

hiddeco opened the pull request:
This PR decouples the synchronization of charts from git sources from the synchronization of `HelmRelease` resources to Helm releases and/or cluster resources.

The synchronization of charts from Git sources has been rewritten so that we only do bookkeeping of git references for `HelmRelease` resources, but no longer maintain a working directory, this copy is instead made on request based on the references we have in our books at that moment.

The release synchronization is moved to the `release` package, and has been rewritten so that implementers no longer have to decide what actions they want to perform, they can simply call `Sync` (install or upgrade on creation and/or update of `HelmRelease` resource) or `Uninstall` (uninstall on delete of `HelmRelease` resource).

Fixes #98

hiddeco committed:
Decouple the sync of git chart sources

hiddeco committed:
Decouple release synchronization from chart sync

hiddeco committed:
Fix: actually compare the mirror name for the HR

hiddeco committed:
Release: remove failed initial release

To make Helm v3 a first class citizen, it would be better to mimic
the `--atomic` behaviour from Helm v3 in the v2 client code, so that
the release is removed automagically if the installation fails.

This is however outside the scope of the PR this commit is part of.

2opremio commented on pkg/chartsync/download.go:
```diff
@@ -49,6 +35,20 @@ func ensureChartFetched(base string, source *helmfluxv1.RepoChartSource) (string
+	// to them that is based on the URL.
+	repoPath := filepath.Join(base, base64.URLEncoding.EncodeToString([]byte(source.CleanRepoURL())))
+	if err := os.MkdirAll(repoPath, 00750); err != nil {
+		panic(err)
```
I know you just moved this but I am generally against panics. It's OK to exit, but with a friendlier error than a stack trace

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+	if s, _ := c.syncGitChartSource(repo, hr); s != nil {
+		return s, true
+	}
+	return nil, false
```
You can replace all this by:
```go
s, _ := c.syncGitChartSource(repo, hr)
return s, s != nil 
```

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+
+	// Check if we have a source in store and return if it still equals
+	// to what is configured in the source.
+	c.sourcesMu.Lock()
```
You may want to consider using a `RWMutex` and `RLock()`ing here instead of `Lock()`ing

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+
+	wg.Add(1)
+	go func() {
+		defer runtime.HandleCrash()
```
Is this really necessary?

hiddeco replied:
I do not think so, I think it is there for historical reasons, as the `panic` you commented on earlier in a previous version of the operator could have been triggered by this loop.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+			select {
+			case changed := <-c.mirrors.Changes():
+				for mirrorName := range changed {
+					mirror, ok := c.mirrors.Get(mirrorName)
```
This function is pretty large, I would extract the body of the for loop into a separate method (e.g. `processChangedMirror()`)

2opremio replied:
Maybe it's better to only factor out the `ok == true` part of the body.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+	if !ok {
+		// No source found in store;
+		// create the boilerplate of a new one.
+		s = &GitChartSource{Mutex: sync.Mutex{}, Mirror: mirrorName, Remote: source.GitURL}
```
I don't think it's necessary to initialize the mutex

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+// has seen commits for paths we are interested in. It returns the
+// synchronized source or nil and a boolean indicating if the source
+// was updated.
+func (c *GitChartSourceSync) syncGitChartSource(repo *git.Repo, hr *v1.HelmRelease) (*GitChartSource, bool) {
```
This function is a bit too long. Anything you can factor out from here?

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+		defer oldExport.Clean()
+	}
+
+	defer func() {
```
Any reason for not simply doing this at the end of the function instead of using `defer`?

hiddeco replied:
I want to set the condition on the `HelmRelease` _after_ the source has become available so that the timestamp of the condition update is technically correct. Which isn't the case as long as this method holds the lock.

2opremio replied:
We are probably talking about a sub-millisecond time difference, is that really worth it? At least I think we should add a comment.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+		c.sources[hr.ResourceID().String()] = s
+	}
+	// Acquire source lock and unlock sources lock.
+	s.Lock()
```
Locking the entry while holding map lock means that doing the opposite (locking the map while holding the lock of an entry) can cause a deadlock. I haven't been able to see that happening in the code, but it's something to keep in mind.

2opremio commented on pkg/release/release.go:
```diff
@@ -1,556 +1,342 @@
+		// Lock the resource so that it does not get updated by
+		// the chart sync process.
+		s.Lock()
+		defer s.Unlock()
```
Do you need to hold the lock throughout the execution of the rest of function?

hiddeco replied:
Yes, as otherwise a signal could replace the working clone and clean up the old one while the release has not happened yet. Causing the path to the chart to no longer exist.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,407 @@
+// GitChartSource holds the source references to a Helm chart in git,
+// and a minimal working clone.
+type GitChartSource struct {
+	sync.Mutex
```
Exposing the mutex externally makes it hard to reason about concurrency issues. Also it's not obvious (to me at least) when should one hold the lock (at least without looking at the implementation of `GitChartSourceSync`). During any action on Export? 

Maybe we can provide the same concurrency guarantees through a more intuitive/encapsulated abstraction. For instance, is it possible to provide the same guarantees by providing methods for `GitChartSource` instead?

hiddeco replied:
One should hold the lock for the duration one works with the source, as the `Export` has a relationship with all other fields in the structure. Providing multiple methods is thus not an option.

The only way I can think of encapsulating this is to create a callback method, which acquires the lock and defers the unlock. I may however be short sighted in options/possibilities.

hiddeco committed:
Release: restore Prometheus metric

The previous metric kept track of the time it would take to perform
a Helm action, and some other actions around the synchronization of
the `HelmRelease` resource to the cluster, but would not give a real
insight in the total amount of time it would take to process a
`HelmRelease` resource.

This commit slightly tweaks the metric, so that it now measures the
total amount of time it takes to synchronize a single `HelmRelease`
resource to the cluster, and drops labels that due to this change
have become irrelevant.

hiddeco committed:
Chartsync: err on chart path creation failure

hiddeco committed:
Chartsync: use RWMutex for git sources

hiddeco committed:
Chartsync: various small improvements

* remove obsolete `runtime.HandleCrash()` which was still included
  because of historical reasons
* remove unnecessary mutex initialization
* flatten return into something more Golang idiomatic

hiddeco committed:
Chartsync: factor out mirror change processing

Plus various alignments around naming things.

hiddeco committed:
Release: only set `atomic` option during install

If the atomic option is set during upgrades, Helm v3 will automatically
rollback the release if it fails, which is not something we are after
as this can be dangerous for e.g. charts that make use of `StatefulSet`
resources.

During installation however we _do_ want to use this flag, as this
automatically removes a failed release for us, so that we do not have
to perform or implement this logic ourselves.

hiddeco committed:
Chartsync: reduce complexity of Git chart sync

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+// indicates whether the repo was already present (`true` if so,
+// `false` otherwise).
+func (c *GitChartSync) maybeMirror(mirrorName string, remote string) bool {
+	ok := c.mirrors.Mirror(
```
Maybe this isn't a new problem but, I only see a `StopOne()` counterpart to this `Mirror()` call in `Delete()` which seems to only be invoked when a `HelmRelease()` is deleted.

Doesn't this mean that there will be a mirror leak if the `git` entry of a `HelmRelease` is changed (or removed)?

2opremio replied:
Additionally, `Delete()` doesn't help once the `git` field is modified, since it uses `helmReleasesForMirror()` which relies in live data, and not the mirrors created by the `HelmRelease()`

2opremio replied:
If we don't want to track changes on the`git` field explicitly, I think this can be fixed by periodically comparing the output of `helmReleasesForMirror()` on `c.mirrors`. I don't think it's ideal though.

2opremio replied:
Now I see that `forHelmRelease()` is used in `get()` to take into account that  the repo of `HelmRelease`s  can change with respect to what's book-kept in `GitChartSync`, but I don't think we do anything to correct divergences.

hiddeco replied:
> Maybe this isn't a new problem but, I only see a `StopOne()` counterpart to this `Mirror()` call in `Delete()` which seems to only be invoked when a `HelmRelease()` is deleted.

The current `master` version does not stop any mirrors, they continue to live forever. I will let your comments simmer in my head for a while.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+	// Attempt to get the source from store.
+	source, ok := c.sources[hr.ResourceID().String()]
+	if ok {
+		// Remove the in store source.
```
Remove the ... release?

hiddeco replied:
We are not deleting the release here but our indexation / bookkeeping of the git source. I think the comment does not add much now that the code has been reduced.

2opremio replied:
Ah, you meant the _in-store_ source (sorry, I should had understood)

hiddeco replied:
Yes, I already thought the missing hyphen was the actual issue.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+		defer func() {
+			c.mirrors.StopAllAndWait()
+			wg.Done()
+		}()
```
Why not place this code directly in the `case <-stopCh:` instead of using a deferred function call?

hiddeco replied:
One of the few original lines from the original code.

Historically looking back, I think this is because the previous `panic` possibility, which would still make the `defer` happen (but would not reach the `stopCh`).

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+
+// SyncMirrors instructs all git mirrors to sync from their respective
+// upstreams.
+func (c *GitChartSync) SyncMirrors() {
```
This seems to be needed because of https://github.com/fluxcd/helm-operator/commit/c20e5c84c2e4580c3ab5eba157250c012a1ae9c3 ,  but I cannot find a matching PR. Would you mind clarifying why the http endpoint was added?

hiddeco replied:
It is the operator’s equivalent of `fluxctl sync`, but via an HTTP endpoint, to give users the option to instantly trigger the refresh from git mirrors from e.g. CI, instead of waiting for the poll interval to kick in.

hiddeco replied:
Original PR: https://github.com/fluxcd/flux/pull/1776

2opremio replied:
Of course, I forgot this repo was split out recently, sorry.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+
+// GitChartDir holds a `git.Export` of a chart repository,
+// and the commit hash of the HEAD of this export.
+type GitChartDir struct {
```
I was going to ask for a name change (something like `GitChartWorkingDir`) but, after seeing its use I think its overkill to have an explicit structure. I would recommend getting rid of it and returning an `Export` and the `Head` explicitly (`ChartPath` can be implemented by the consumer).

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+
+// ChartDir returns a newly exported `GitChartDir` for the given `v1.HelmRelease`
+// at the latest HEAD, or an error.
+func (c *GitChartSync) ChartDir(hr *v1.HelmRelease) (GitChartDir, error) {
```
I would rename this to something with a verb (since it requires some work) and more descriptive. Something like `GetChartSnapshotDir()` or `GetMirrorCopy()` or `TakeChartMirrorSnapshot()` or `TakeChartSnapshot()`

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+	return sourceRef{}, false
+}
+
+func (c *GitChartSync) store(hr *v1.HelmRelease, s sourceRef) {
```
This is only used once, so I would simply inline it.

2opremio replied:
or at least rename it to `storeSource`

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+	}
+}
+
+func (c *GitChartSync) get(hr *v1.HelmRelease) (sourceRef, bool) {
```
This is only used once, so I would simply inline it.

2opremio replied:
or at least rename it to `getSource`

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+	mirrors *git.Mirrors
+
+	sourcesMu sync.RWMutex
+	sources   map[string]sourceRef
```
I think it's worth adding a comment about the map being indexed by the `ResourceID` of `HelmReleases` and probably also renaming it to `releaseSources` or `releaseSourcesByID`

hiddeco replied:
I felt that changing the name to `releaseSourcesByID` made the comment itself obsolete.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+// record). In case of failure it returns an error.
+func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Repo) (sourceRef, bool, error) {
+	source := hr.Spec.GitChartSource
+	if source == nil {
```
It's strange that the method needs to check for the `GitChartSource` and also provide a `mirrorName`, which should be a subproduct of the former).

2opremio replied:
I would change the interface to `sync(hrID resource.ID, mirrorName string, repo *git.Repo) ` 

But, that's not currently possible, because `sync` also consistency-checks that the mirror we have for the Helm release (in `get()`s invocation to `forHelmRelease()`).

However, I don't think that consistency-checking is the direct responsibility of `sync()` (or at least it makes it harder to understand what's going on since it's buried when invoking `get()`).

I think that we should take car of divergences explicitly, while also garbage collecting orphan mirrors (i.e. https://github.com/fluxcd/helm-operator/pull/99/files#r347533178 )

hiddeco replied:
There is another reason this is not possible, without the `v1.HelmRelease` or a copy of the `v1.GitChartSource`, it is not possible to detect changes as done on L286, nor is it possible to construct a new `sourceRef` as done on L265.

2opremio commented on pkg/chartsync/git.go:
```diff
@@ -0,0 +1,337 @@
+// helmReleasesForMirror returns a slice of `HelmRelease`s that make
+// use of the given mirror.
+func (c *GitChartSync) helmReleasesForMirror(mirror string) ([]*v1.HelmRelease, error) {
+	hrs, err := c.lister.List(labels.Everything())
```
Isn't it a bit overkill to list all the helm releases every time? I don't know the operator well, but I presume we use a `HelmRelease` informer, which we can use to cache the mirrors and keep them updated.

2opremio replied:
We could use the same cache to make sure that `c.sources` are up to date, allowing to garbage collect mirrors and (maybe) allowing to remove the consistency check in `get()`

hiddeco replied:
This is a `HelmReleaseLister`, which is nothing more then a named `cache.Indexer`, we basically loop over in-memory `HelmRelease` resources at close-to-zero cost.

hiddeco replied:
(I can also give you some historical insights and tell you that before my previous refactor round, we used the client for this type of operations 🤫).

2opremio replied:
> we basically loop over in-memory HelmRelease resources at close-to-zero cost.

Ah, I see, I thought it involved calling the client 👍

2opremio commented:
I think that the implementation has been drastically improved and looks much cleaner 🏅 , particularly when it comes to concurrency

However, this has allowed me to focus on other parts, particularly the consistency of `git references from`HelmReleases` which I think can be improved.

hiddeco committed:
Chartsync: be explicit about map contents

hiddeco committed:
Chartsync: rename `ChartDir` -> `GetMirrorCopy`

Plus remove the explicit structure and instead return the `*git.Export`
and HEAD reference string explicitly.

hiddeco committed:
Chartsync: move `defer` to `<-stopCh`

hiddeco committed:
Chartsync: inline `get` and `store` methods

hiddeco commented:
@fons I would like to propose to merge this _as is_, and tweak the bits around garbage collection and/or the imperfect clean up of mirrors at a later stage. This PR is holding back the developments around Helm v3 - which I would like to focus on the rest of this week, and the current state of the PR is arguably already ten times better as what it was.

We can work out the missing piece of the puzzle in an issue, as we both do not seem to have a direct answer to the question what the best way forward is for this given the triangular relationship of `HelmRelease` -> `sourceRef` <- `mirror(Signal)`. I also think having some time off from this code, while focusing on other elements of the operator, will probably lead to a solution faster than being stuck on minor details.

2opremio commented:
Sounds good! Can you create a ticket for it?

hiddeco committed:
Chartsync: garbage collect mirror without refs

This commit brings the leakage of mirrors close to zero, as the
mirroring will be stopped if there are no references to it from
`HelmRelease` resources on the first signal received.
